buildSrc/src/main/java/org/elasticsearch/gradle/test/RestTestRunnerTask.java

```diff
@@ -0,0 +1,39 @@
+package org.elasticsearch.gradle.test;
+
+import org.elasticsearch.gradle.testclusters.ElasticsearchCluster;
+import org.gradle.api.tasks.CacheableTask;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.testing.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import static org.elasticsearch.gradle.Distribution.INTEG_TEST;
+
+/**
+ * Customized version of Gradle {@link Test} task which tracks a collection of {@link ElasticsearchCluster} as a task input. We must do this
+ * as a custom task type because the current {@link org.gradle.api.tasks.TaskInputs} runtime API does not have a way to register
+ * {@link Nested} inputs.
+ */
+@CacheableTask
+public class RestTestRunnerTask extends Test {
+
+    private Collection<ElasticsearchCluster> clusters = new ArrayList<>();
```

Comment:
That get's a little awkward as then the task would have to keep a reference back to the extension or similar. I think that's bad practice. It's the job of the _plugin_ to do this wiring. The task should not have any knowledge of the test clusters plugin.

Reply:
The  task already cares about the clusters so does know a lot about the  plugin.
We talked about being able to use clusters with specific types of tasks only in the past with @rjernst and agreed it's a good idea, it just never got done, so another way to think about it is to consider the task part of testClusters so that it's the only one that can use them. `useCluster` could become a method on this task instead of an extension, possibly saving us a bit at config time. All the house-keeping (claims, running cluster etc) could live in an extension so the task can reference it via the `Project`.  That would make the plugin code more readable too. We have a handful of  
`Test` tasks using testclusters but those could and should be `RestTestRunnerTasks`.

Reply:
It's still not clear exactly what the proposal here is. If the suggestion is to move logic that exists currently in `TestClustersPlugin` into this task then frankly that's a flat 👎 for me. Tasks should be dumb. They should not contain complex logic for managing state or inputs aside from their intended purpose. That code should and _is_ in the plugin. If that means exposing more properties on the task for the plugin to talk to it then so be it. At some point this becomes a philosophical argument but I'm going to remain stubborn here. `RestIntegTestTask` is the extreme example of what I'm talking about. A bunch of complex build logic that got sucked up into a task instead of a plugin and it's a nightmare.

I simply don't see the issue with this method. This task implementation is dead simple. All it does is expose a new bean property, and that's its only intended purpose.

Reply:
Going back and reading your comment I think I might better understand. If the idea is "if you want to use a test cluster you have to configure a `RestTestRunnerTask`" I think that's moving in the right direction. In the end what we want is _something_ to replace `RestIntegTestTask`, that is, some central configurable DSL element that setups up all this stuff for us (test task, dependencies, test cluster, etc). What thing thing might be I think is still up for a discussion, but it needs to be higher level than a task as eventually I think we'd like this thing to potentially create source sets, etc. I end goal here is that for REST tests a user should simply need to dump YAML files in a folder, tell us what cluster configuration they need and be on their way.

In terms of this PR, `RestTestRunnerTask` is meant to solve a very specific problem which is to track an additional input. There's no implied intention for this task to solve the problems I state above, those need to be addressed elsewhere. I think it's worth discussing this more in our weekly tomorrow though as I do think we share the same high-level goals here.